graphql-api/src/queries/helpers/json-cache.ts

```diff
@@ -0,0 +1,190 @@
+    const jsonString = JSON.stringify(obj);
+
+    if (useCompression) {
+      const compressedData = await gzipAsync(jsonString)
+      await fs.writeFile(path, compressedData)
+    } else {
+      await fs.writeFile(path, jsonString, 'utf8');
+    }
+  } catch (err) {
+    logger.warn(`Error writing JSON to file: ${err}`);
+    throw err;
+  }
+}
+
+export async function readJsonFromFile<T>(path: string, useCompression = false): Promise<T> {
+  try {
+    if (useCompression) {
+      const data = await fs.readFile(path);
+      const decompressedContents = await gunzipAsync(data);
+      return JSON.parse(decompressedContents.toString('utf8')) as T
+    }
+    const data = await fs.readFile(path, "utf8");
+    return JSON.parse(data) as T;
+  } catch (err) {
+    logger.warn(`Error reading JSON to file: ${err}`);
+    throw err;
+  }
+}
+
+async function fileExists(path: string): Promise<boolean> {
+  try {
+    await fs.access(path);
+    return true;
+  } catch (err) {
+    return false;
+  }
+}
+
+
+export async function writeJsonToFileGcs(obj: any, bucket: Bucket, path: string, useCompression = false): Promise<void> {
+
+  const file = bucket.file(path)
+  const data = JSON.stringify(obj)
+
+  if (useCompression) {
+    const compressedData = await gzipAsync(data);
+    await file.save(compressedData)
+  } else {
+    await file.save(data)
+  }
+}
+
+export async function readJsonFromFileGcs<T>(path: string, bucket: Bucket, useCompression = false): Promise<T> {
+  const file = bucket.file(path)
+  const [maybeCompressedContents] = await file.download()
+  if (useCompression) {
+    const contents = await gunzipAsync(maybeCompressedContents);
+    return JSON.parse(contents.toString('utf-8'))
+  }
+  return JSON.parse(maybeCompressedContents.toString('utf-8'))
+}
+
+async function fileExistsGcs(path: string, bucket: Bucket): Promise<boolean> {
+  const file = bucket.file(path)
+  const [exists] = await file.exists()
+  logger.info(`Key ${path} exists in cache`)
+  return exists
+}
+
+function extractBucketName(gcsPath: string): string | null {
+  const regex = /^gs:\/\/([^/]+)\/?.*$/;
+  const match = gcsPath.match(regex);
+
+  if (match && match[1]) {
+    return match[1];
+  }
+  return null
+}
+
+function extractGcsPath(gcsUrl: string): string | null {
+  const regex = /^gs:\/\/[^/]+\/(.+)$/;
+  const match = gcsUrl.match(regex);
+
+  if (match && match[1]) {
+    return match[1];
+  }
+  return null
+}
+
+export class JsonCache<T> {
+  private json_cache_path: string
+
+  private bucketName: string | null
+
+  private bucket?: Bucket
+
+  private useCompression = false
+
+  filePathFromCacheKey(key: string): string {
+    const gzipExt = this.useCompression ? ".gz" : ""
+    const fileName = `${key.replace(/:/g, "-")}.json${gzipExt}`
+    return join(this.json_cache_path, fileName)
+  }
+
+  constructor(json_cache_path: string, compression = false) {
+    this.json_cache_path = json_cache_path
```

Comment:
This is my limited typescript experience speaking -- but I think I'm reading that we fall back to a local filepath if we're _not_ using a bucket? What happens if the json cache path variable is empty? Is it writing files in the root (`/`) directory of the container?

Reply:
If `gs://` is in the file path, it will use a bucket. If not, it will use a local file path. That could probably be more clear.

Since the constructor only accepts a string (and not null/undefined), it will never be empty.

Reply:
I agree about the need for docs of some kind.

Reply:
OK, yeah, i'd say some docs on how to determine if the JSON cache needs invalidation and the preferred procedure to clear it would be good, but i don't feel like that needs to block merging the code

https://github.com/broadinstitute/gnomad-browser/issues/1331